[jit] Support named tuple in tracing

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29751 [jit] Support named tuple in tracing**

Summary:
Our named tuple in tracing is not feature complete, it only support
making named tuple as output and will lose the names once we did that.
This PR make sure we could support named tuple as input and retain the
names during tracing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18491297](https://our.internmc.facebook.com/intern/diff/D18491297)

wanchaol committed:
[jit] More support named tuple in tracing

Summary:
Our named tuple in tracing is not feature complete, it only support
making named tuple as output and will lose the names once we did that.
This PR make sure we could support named tuple as input and retain the
names during tracing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] More support named tuple in tracing"


Summary:
Our named tuple in tracing is not feature complete, it only support
making named tuple as output and will lose the names once we did that.
This PR make sure we could support named tuple as input and retain the
names during tracing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

bddppq commented on torch/csrc/jit/tracer.cpp:
```diff
@@ -203,13 +203,21 @@ Value* TracingState::getOutput(const IValue& iv, size_t i) {
+    TupleTypePtr tuple_type = iv.type()->cast<TupleType>();
+    Node* tuple_node = nullptr;
+    if (tuple_type->name() && tuple_type->schema()) {
+      tuple_node = graph->createTuple(
```
could you add a comment here saying only named tuple has schema

bddppq reviewed (commented):
Thanks!

wanchaol committed:
Update on "[jit] More support named tuple in tracing"


Summary:
Our named tuple in tracing was not feature complete, it only supported
making named tuple as output and will lose the names once we did that.
This PR make sure we could support named tuple as input and retain the
names during tracing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18491297](https://our.internmc.facebook.com/intern/diff/D18491297)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] More support named tuple in tracing"


Summary:
Our named tuple in tracing was not feature complete, it only supported
making named tuple as output and will lose the names once we did that.
This PR make sure we could support named tuple as input and retain the
names during tracing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18491297](https://our.internmc.facebook.com/intern/diff/D18491297)

[ghstack-poisoned]

wanchaol commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -170,6 +170,14 @@ inline InferredType tryToInferContainerType(py::handle input) {
     }
+    if (PyObject_HasAttrString(input.ptr(), "_fields")) {
+      // If this is a NamedTuple type, fetch the names correspondingly.
+      auto qualifiedName = c10::QualifiedName(py::cast<std::string>(
```
unfortunately we could not call `get_named_tuple_property` function here because it will try to resolve the python type annotation again where it does not exist in tracing (tracing does not have compilation unit to hold registered types).

zdevito replied:
This code is going to create a new NamedTuple type for every input value, even if it has the same type as another named tuple object in the system. This will cause future bugs. E.g. a scripted network that contains a traced sub-network that contains named tuples types. The traced sub-network will not be able to ever invoke the trace because the named tuple will have been resolved to a different type. Tracing should re-use the same infrastructure that named tuples and classes use in scripting. It can use the python_compilation_unit that all other python code uses.

zdevito commented on torch/csrc/jit/tracer.cpp:
```diff
@@ -203,13 +203,25 @@ Value* TracingState::getOutput(const IValue& iv, size_t i) {
-    auto tuple_node = graph->createTuple(
-        fmap(tuple, [&](const IValue& ival) { return getOutput(ival, i); }));
+    TupleTypePtr tuple_type = iv.type()->cast<TupleType>();
+    Node* tuple_node = nullptr;
```
No need for this if statement, the type can be passed unconditionally. createTuple will work if the type passed is not a named tuple as well.

wanchaol replied:
There's an assert that we either not passing tuple type or the tuple type is a named tuple https://github.com/pytorch/pytorch/blob/309b28ee3a4ba8f916b6a377fe0d3d11d87f2592/torch/csrc/jit/ir.cpp#L1448

I will remove that assert then.

zdevito reviewed (commented):
Thanks for work on this -- we need to make sure named tuple creation goes through the same pathway as script for compatibility reasons, then this will be good to go.

wanchaol committed:
Update on "[jit] More support named tuple in tracing"


Summary:
Our named tuple in tracing was not feature complete, it only supported
making named tuple as output and will lose the names once we did that.
This PR make sure we could support named tuple as input and retain the
names during tracing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18491297](https://our.internmc.facebook.com/intern/diff/D18491297)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] More support named tuple in tracing"


Summary:
Our named tuple in tracing was not feature complete, it only supported
making named tuple as output and will lose the names once we did that.
This PR make sure we could support named tuple as input and retain the
names during tracing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18491297](https://our.internmc.facebook.com/intern/diff/D18491297)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] More support named tuple in tracing"


Summary:
Our named tuple in tracing was not feature complete, it only supported
making named tuple as output and will lose the names once we did that.
This PR make sure we could support named tuple as input and retain the
names during tracing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18491297](https://our.internmc.facebook.com/intern/diff/D18491297)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] More support named tuple in tracing"


Summary:
Our named tuple in tracing was not feature complete, it only supported
making named tuple as output and will lose the names once we did that.
This PR make sure we could support named tuple as input and retain the
names during tracing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18491297](https://our.internmc.facebook.com/intern/diff/D18491297)

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -170,6 +170,25 @@ inline InferredType tryToInferContainerType(py::handle input) {
+        // If the python_cu already contains the type, just return it
+        return InferredType(named_tuple_type);
+      } else {
+        // Otherwise fetch the names correspondingly and create a new NamedTuple type
```
This code is a reimplementation of the logic in PythonResolver::resolveType. They need to unified into a single function can called from both location. For instance the check for _fields here is just a reimplementation of isNamedTupleClass in python resolver.

wanchaol replied:
I actually tried merging these two together, there're some problems: 1. We don't expose SourceRange to `tryToInferType` so we couldn't throw ErrorReport like resolveType. (Maybe we can get the python interpreter SourceRange here) 2. we could not call get_named_tuple_property function here because it will try to resolve the python type annotation again (with `ann_to_type`) where it does not exist in tracing (we need to pass in `rcb` to `ann_to_type` to know lazily created types from python annotations).

wanchaol committed:
properly share type creation on "[jit] More support named tuple in tracing"


Summary:
Our named tuple in tracing was not feature complete, it only supported
making named tuple as output and will lose the names once we did that.
This PR make sure we could support named tuple as input and retain the
names during tracing.


Fixes https://github.com/pytorch/pytorch/issues/23355

Differential Revision: [D18491297](https://our.internmc.facebook.com/intern/diff/D18491297)

[ghstack-poisoned]

wanchaol commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -170,6 +234,18 @@ inline InferredType tryToInferContainerType(py::handle input) {
+        input_type,
+        qualifiedName,
+        element_types,
+        tracer::getPythonInterpreterSourceRange());
```
I am not sure `tracer::getPythonInterpreterSourceRange()` should be the way to get the source range here, but I think it's acceptable since we always have python interpreter here.

wanchaol committed:
minor fix on "[jit] More support named tuple in tracing"


Summary:
Our named tuple in tracing was not feature complete, it only supported
making named tuple as output and will lose the names once we did that.
This PR make sure we could support named tuple as input and retain the
names during tracing.


Fixes https://github.com/pytorch/pytorch/issues/23355

Differential Revision: [D18491297](https://our.internmc.facebook.com/intern/diff/D18491297)

[ghstack-poisoned]

wanchaol commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -76,6 +77,69 @@ inline TypedIValue toDictKeyIValue(py::handle key) {
+
+  if (auto type = get_python_cu()->get_type(qualName)) {
+    TORCH_CHECK(
+        type->isSubtypeOf(named_type) || named_type->isSubtypeOf(type),
```
There's another problem when trying to share the code and the Type with scripting/tracing. During tracing we infer the TensorTypes from the input and create the TensorTypes with sizes. But Scripting we only create plain TensorTypes. For the NamedTuple case, it's a container type, and NamedTupleType created from tracing might be `NamedTuple(Tensor(sizes1), Tensor(size2))`, but NamedTupleType created from scripting is just `NamedTuple(Tensor, Tensor)`. 

There's convariance problem on the subtyping relationship between these two type. I changed the subtype check here to check both conditions to ensure there's at least one pass. But ideally we need to define the behavior of it correctly. 

One idea is that when we creating the NamedTuple type from tracing, we pass the unshaped types to the type creation, that way we should the exact types between tracing/scripting. But this might affect the profiling graph as we basically erase the input tensor shapes from the named tuple.

wanchaol replied:
this get fixed by https://github.com/pytorch/pytorch/pull/30706

wanchaol committed:
Update on "[jit] Support named tuple in tracing"


Summary:
Our named tuple in tracing is not feature complete, it only support
making named tuple as output and will lose the names once we did that.
This PR make sure we could support named tuple as input and retain the
names during tracing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18491297](https://our.internmc.facebook.com/intern/diff/D18491297)

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/script/compilation_unit.h:
```diff
@@ -186,12 +186,11 @@ struct TORCH_API CompilationUnit {
+      return nullptr;
     }
-    return nullptr;
+    return type->expect<TupleType>();
```
This function seems suspicious because it can error if the type exists but is not a named tuple, but it returns nullptr (and does not error) if no type with the name is found.

wanchaol replied:
hmm yeah this follows the same logic as `get_class` or `get_interface`, and these methods are mainly used by compilation_unit api, do you think we should throw error in those cases if no type if found?

zdevito commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -77,6 +78,69 @@ inline TypedIValue toDictKeyIValue(py::handle key) {
+inline TypePtr getOrCreateNamedTupleType(
+  py::object obj,
+  const c10::QualifiedName& qualName,
+  c10::optional<std::vector<TypePtr>> element_types,
```
I don't understand where these can be optional.

wanchaol replied:
this is because we want to share the named tuple creation with tracing and scripting. 

* tracing: we will infer the inner types of the container type according to inputs, and `element_types` are the inner types that the tracer inferred. 

* scripting: we couldn't always infer the inner types as tracing since we might not have input from python during scripting. What we did here is we use the type from annotations `annotations.ann_to_type` of the named tuple class.

zdevito commented on torch/jit/__init__.py:
```diff
@@ -2081,17 +2081,22 @@ def _check_directly_compile_overloaded(obj):
 Error.__qualname__ = "Error"
 
-def _get_named_tuple_properties(obj):
+def _get_named_tuple_properties(obj, need_types=True):
```
Booleans that change the return type of a function are not a great way to organize code. Can these be made into two functions.

zdevito commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -171,6 +235,18 @@ inline InferredType tryToInferContainerType(py::handle input) {
+      c10::TypePtr named_type = getOrCreateNamedTupleType(
+          input_type,
+          qualifiedName,
+          element_types,
```
what happens if the same named tuple is used with different types into two different places in the trace?

wanchaol replied:
same namedtuple should have a single named tuple type in the python cu because we drop the second one if the first one is a subtype of the second one.

zdevito reviewed (commented):
There seem to be some design issues here. In particular I'd like to get more clarity on how this is expected to behave if the same named tuple appears as multiple inputs or outputs. And what happens when a named tuple with no type annotations is used in two places which trace to different types. Once the desired behavior is more clear, it will be easier to review the implementation.

wanchaol commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -77,6 +78,69 @@ inline TypedIValue toDictKeyIValue(py::handle key) {
+    named_type = TupleType::createNamed(qualName, fields, annotations);
+  }
+
+  if (auto type = get_python_cu()->get_type(qualName)) {
```
This is copied from init.cpp, where I think the logic is if we use the same named tuple in the trace:
* first time it will create the named tuple type (`named_type`) as normal.
* second time we hit the same named tuple type, we first still create the `named_type`, then we check if the python_cu already contains `type` that is subtype of `named_type`, if we do, then we return `type`, and drop the newly created `named_type`

jamesr66a replied:
It looks like @suo added this particular check: https://github.com/pytorch/pytorch/pull/23798

NamedTuple types are unique'd via QualifiedName. Does this answer your question @zdevito

wanchaol committed:
Update on "[jit] Support named tuple in tracing"


Summary:
Our named tuple in tracing is not feature complete, it only support
making named tuple as output and will lose the names once we did that.
This PR make sure we could support named tuple as input and retain the
names during tracing.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18491297](https://our.internmc.facebook.com/intern/diff/D18491297)

[ghstack-poisoned]

thuyen commented:
It's been sometime, do we have a plan for this and #27743 in the near future?
